router/core/operation_processor.go

```diff
@@ -919,37 +926,71 @@ func (o *OperationKit) Validate(skipLoader bool) (cacheHit bool, err error) {
 	return
 }
 
-// ValidateQueryDepth validates that the operation query depth isn't greater than the max query depth.
-func (o *OperationKit) ValidateQueryDepth(maxQueryDepth int, operation, definition *ast.Document) (bool, int, error) {
-	if o.cache != nil && o.cache.queryDepthCache != nil {
-		depth, cacheHit := o.cache.queryDepthCache.Get(o.parsedOperation.ID)
-		if cacheHit {
-			valid := depth <= maxQueryDepth
-			if !valid {
-				return cacheHit, depth, &httpGraphqlError{
-					message:    fmt.Sprintf("The query depth %d exceeds the max query depth allowed (%d)", depth, maxQueryDepth),
-					statusCode: http.StatusBadRequest,
-				}
-			}
-			return cacheHit, depth, nil
+// ValidateQueryComplexity validates that the query complexity is within the limits set in the configuration
+func (o *OperationKit) ValidateQueryComplexity(complexityLimitConfig *config.ComplexityLimits, operation, definition *ast.Document, isPersisted bool) (bool, ComplexityCacheEntry, error) {
+	if o.cache != nil && o.cache.complexityCache != nil {
+		if cachedComplexity, ok := o.cache.complexityCache.Get(o.parsedOperation.ID); ok {
+			return ok, cachedComplexity, o.runComplexityComparisons(complexityLimitConfig, cachedComplexity, isPersisted)
 		}
 	}
 
 	report := operationreport.Report{}
-	globalComplexityResult, _ := operation_complexity.CalculateOperationComplexity(operation, definition, &report)
-	valid := globalComplexityResult.Depth <= maxQueryDepth
+	globalComplexityResult, rootFieldStats := operation_complexity.CalculateOperationComplexity(operation, definition, &report)
+	cacheResult := ComplexityCacheEntry{
+		Depth:       globalComplexityResult.Depth,
+		TotalFields: globalComplexityResult.NodeCount,
+	}
+	for _, entry := range rootFieldStats {
+		if entry.Alias == "" {
+			cacheResult.RootFields += 1
+		} else {
+			cacheResult.RootFieldAliases += 1
+		}
+	}
 
-	if o.cache != nil && o.cache.queryDepthCache != nil {
-		o.cache.queryDepthCache.Set(o.parsedOperation.ID, globalComplexityResult.Depth, 1)
+	if o.cache != nil && o.cache.complexityCache != nil {
+		o.cache.complexityCache.Set(o.parsedOperation.ID, cacheResult, 1)
 	}
 
-	if !valid {
-		return false, globalComplexityResult.Depth, &httpGraphqlError{
-			message:    fmt.Sprintf("The query depth %d exceeds the max query depth allowed (%d)", globalComplexityResult.Depth, maxQueryDepth),
-			statusCode: http.StatusBadRequest,
+	return false, cacheResult, o.runComplexityComparisons(complexityLimitConfig, cacheResult, isPersisted)
+}
+
+func (o *OperationKit) runComplexityComparisons(complexityLimitConfig *config.ComplexityLimits, cachedComplexity ComplexityCacheEntry, isPersisted bool) error {
+	type comp struct {
+		field        int
+		cachedField  int
+		errorMessage string
+	}
+
+	testComparisons := []comp{}
+	if complexityLimitConfig.Depth != nil && complexityLimitConfig.Depth.Enabled && (!isPersisted || !complexityLimitConfig.Depth.IgnorePersistedOperations) {
```

Comment:
maybe add a methods on a config smth like

```
func (c *config) shouldAssertDept(isPersisted) {
  return c.Depth != nil && ...
}
```
It will be much more readable